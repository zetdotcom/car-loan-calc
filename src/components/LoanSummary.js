import React from 'react';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import { DATE_FORMAT } from 'constants/common';
import { pounds } from 'utils';
import { Heading2, Heading3 } from './ui/Headings';
import styled from 'styled-components';

const PaymentScheduleList = styled.ul`
	display: flex;
	flex-direction: column;
	padding: 0;
	align-items: center;
`;

const PaymentScheduleListItem = styled.li`
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	border: 1px solid gray;
	background: lightgray;
	border-radius: 4px;
	list-style: none;
	width: 200px;
	margin: 5px;
	& span {
		font-size: 0.8em;
	}
`;

export default function LoanSummary(props) {
	const { monthly, loanTotal, paymentSchedule } = props;

	return (
		<div>
			<Heading2>Loan summary</Heading2>
			<Heading3 data-testid='totalLoanDisplay'>Total loan amount (incl. fees): {pounds(loanTotal)}</Heading3>
			<Heading3 data-testid='monthlyPaymentDisplay'>Monthly payment: {pounds(monthly)}</Heading3>
			<PaymentScheduleList>
				{paymentSchedule.map(item => (
					<PaymentScheduleListItem key={item.date} data-testid='payListItem'>
						<p>
							{format(item.date, DATE_FORMAT)} - {pounds(item.amount)}
						</p>
						<span data-testid='payListSpan'>{item.info}</span>
					</PaymentScheduleListItem>
				))}
			</PaymentScheduleList>
		</div>
	);
}

LoanSummary.propTypes = {};
