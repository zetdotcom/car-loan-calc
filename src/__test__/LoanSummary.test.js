import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import LoanSummary from 'components/LoanSummary';

describe('Loan Summary', () => {
	const defaultProps = {
		monthly: '',
		loanTotal: '',
		paymentSchedule: [],
	};

	const renderComponent = newProps => {
		const props = {
			...defaultProps,
			...newProps,
		};
		const utils = render(<LoanSummary {...props} />);
		const component = <LoanSummary {...props} />;
		return {
			component,
			...utils,
		};
	};

	afterEach(cleanup);

	it('should match snapshot', () => {
		const { asFragment } = renderComponent();
		expect(asFragment()).toMatchSnapshot();
	});
	it('should display correct total and monthly amount and format', () => {
		const { getByTestId, getAllByTestId, rerender } = renderComponent();

		const mockPaymentSchedule = [
			{ date: 1583023189882, amount: 421.3333333333333, info: 'incl. £80 arrangement fee' },
			{ date: 1585697989882, amount: 333.3333333333333, info: '' },
			{ date: 1588289989882, amount: 353.3333333333333, info: 'incl. £20 completion fee' },
		];

		const totalLoan = getByTestId('totalLoanDisplay');
		const monthly = getByTestId('monthlyPaymentDisplay');

		expect(totalLoan).toHaveTextContent('£0.00');
		expect(monthly).toHaveTextContent('£0.00');

		rerender(renderComponent({ loanTotal: 8108, monthly: 100, paymentSchedule: mockPaymentSchedule }).component);
		const payListItem = getAllByTestId('payListItem');
		const payListSpan = getAllByTestId('payListSpan');
		expect(totalLoan).toHaveTextContent('£8,108.00');
		expect(monthly).toHaveTextContent('£100.00');
		expect(payListItem[0]).toBeInTheDocument();
		expect(payListSpan[0]).toBeInTheDocument();
		expect(payListItem[0]).toHaveTextContent('01 Mar 2020 - £421.33');
		expect(payListSpan[0]).toHaveTextContent('incl. £80');
	});
});
