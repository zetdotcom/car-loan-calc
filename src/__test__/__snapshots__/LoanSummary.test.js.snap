// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loan Summary should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="sc-AxirZ mSIYX"
    >
      Loan summary
    </h2>
    <h3
      class="sc-AxiKw duxOnD"
      data-testid="totalLoanDisplay"
    >
      Total loan amount (incl. fees): £0.00
    </h3>
    <h3
      class="sc-AxiKw duxOnD"
      data-testid="monthlyPaymentDisplay"
    >
      Monthly payment: £0.00
    </h3>
    <ul
      class="sc-AxhCb jXGAek"
    />
  </div>
</DocumentFragment>
`;
